### PR TITLE
Add direct Turso database migration

### PR DESCRIPTION
**Pull request by stefan-burke**
## Summary

- Add `deno task migrate:turso`, which asks for the source database, the new Turso name, and any Turso details that are not already in `.env`.
- Download one complete SQLite file, check it, and send that file through Turso's database upload API. The migration does not replay SQL statements one at a time.
- Check that the new name is free before downloading. Remove an incomplete Turso database if setup, upload, or an interrupted run fails.
- Print the new `DB_URL` and `DB_TOKEN` when the migration succeeds, even if temporary-file cleanup then needs operator help.

## Safety

The task hides entered tokens, streams large files with an exact byte length, validates Turso responses before using them, and keeps temporary files owner-only. The app encryption key is not in the database file, so the task reminds the operator to keep using the source `DB_ENCRYPTION_KEY`.

## Testing

- `nix develop -c deno task precommit`
- Turso client mutation test: 66 of 66 changes detected
- Local wire contract verifies the received authorization, byte length, file bytes, and secure upload target
- Snapshot interruption contract verifies that a stalled source sync closes and cleans up

## Summary by CodeRabbit

* **New Features**
  * Added an interactive workflow to migrate a remote database into a new Turso database via `deno task migrate:turso`.
  * Automatically validates the snapshot requirements for Turso upload and uses Turso’s native file upload flow, then outputs the new `DB_URL` and `DB_TOKEN` on success.

* **Documentation**
  * Updated the tasks list and added `TURSO_API_TOKEN`, `TURSO_ORGANIZATION`, and `TURSO_GROUP` to the environment variable docs, including full migration and cleanup steps.

* **Bug Fixes**
  * Improved cancellation/abort handling to stop snapshot sync/verification while still completing required cleanup of incomplete uploads.

* **Tests**
  * Added/extended coverage for migration CLI behavior, uploader validation, snapshot abort safety, and termination-signal handling.

### CODE REVIEW

File: scripts/turso-migration.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import { runDenoScript } from "#scripts/script-runner.ts";
+import { onTerminationSignals } from "#scripts/termination-signals.ts";
+import {
+  uploadTursoDatabaseFile,
+  verifyTursoUploadFile,
+} from "#scripts/turso-migration-file.ts";
+import { runMigrateTursoCli } from "#scripts/turso-migration-lib.ts";
+import { createTursoApi } from "#shared/turso-api.ts";
+
+const fileEnv = await load();
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/turso-migration.ts:16]: Checked this against the pinned `@std/dotenv`: `load()` returns an empty record when the file is absent. `test/scripts/database-snapshot-options.test.ts` directly exercises a missing `.env` path and proves shell credentials are used, and that test passes in the full precommit suite. No code change is needed.

File: src/shared/turso-api.ts [hunk 2]
```diff
@@ -1,98 +1,263 @@
...
- * Rules: 1–63 chars, lowercase letters/numbers/hyphens, no leading/trailing hyphens.
- */
+const TursoDatabaseSchema = v.object({
+  DbId: v.string(),
+  Hostname: v.string(),
+  Name: v.string(),
+});
+const CreateTursoDatabaseSchema = v.object({ database: TursoDatabaseSchema });
+const TursoTokenSchema = v.object({ jwt: v.string() });
+const TursoOrganizationsSchema = v.array(v.object({ slug: v.string() }));
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/turso-api.ts]: Resolved in e79bcdd1. `listOrganizations` now validates the documented `{ organizations: [...] }` envelope and extracts that field. The provider contract test now supplies the real envelope and asserts the returned organization names.

File: scripts/turso-migration-lib.ts [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+): Promise<MigrationOutcome> => {
+  const tempDirectory = await deps.makeTempDir();
+  let credentials: TursoDatabaseCredentials;
+  try {
+    deps.signal.throwIfAborted();
+    deps.stdout("Downloading the source database...");
+    const path = await deps.createSnapshot(
+      { ...source, outputPath: join(tempDirectory, "database.sqlite") },
+      deps.stdout,
+    );
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/turso-migration-lib.ts:124]: Resolved in e79bcdd1. The migration now passes its abort signal into `createDatabaseSnapshot`; snapshot sync races the signal, closes the replica client, and removes temporary files on interruption. `stops a source sync when its signal is aborted` reproduces and locks the stalled-sync case.

File: src/shared/turso-api.ts [hunk 4]
```diff
@@ -1,98 +1,265 @@
...
-    {
-      headers: tursoApiHeaders(),
+/** Build a Turso client around one platform API token. */
+export const createTursoApi = (apiToken: string): TursoApi => {
+  const apiHeaders = (): Record<string, string> =>
+    jsonHeaders({ Authorization: `Bearer ${apiToken}` });
+  const apiFetch = (path: string, init: RequestInit = {}) =>
+    fetchText(`${TURSO_API_BASE}${path}`, {
+      ...init,
+      headers: apiHeaders(),
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/turso-api.ts]: Resolved in f0991a68. `createTursoApi` now accepts an optional `AbortSignal` and forwards it to `fetchText` for all normal operations (list, exists, create, token). `deleteDatabase` deliberately skips the signal so cleanup of an incomplete destination always completes even after an interrupt. The migration passes `deps.signal` to `createApi`. Two contract tests lock both behaviours: `forwards an abort signal to platform API requests` and `does not interrupt cleanup deletes with the abort signal`.

File: scripts/turso-migration-lib.ts [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        api,
+        organization,
+        credentials,
+        error,
+      );
+    }
+  } catch (error) {
+    return await failAfterCleanups(error, [
+      () => deps.removeTempDir(tempDirectory),
+    ]);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/turso-migration-lib.ts]: Resolved in f0991a68. Replaced `failAfterCleanups` with explicit try/catch that formats both the original failure and the cleanup error in one message, including the `tempDirectory` path — mirroring the successful-migration cleanup path. The regression test `reports the original failure and temp directory when cleanup fails` asserts all three pieces appear in the error output.

File: scripts/turso-migration-lib.ts [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    const apiToken =
+      configuredValue(deps, "TURSO_API_TOKEN") ??
+      requiredAnswer(
+        deps.promptSecret("Destination Turso API key:"),
+        "Destination Turso API key",
+      );
+    const source = readSnapshotRequest(
+      { outputPath: "database.sqlite" },
+      (key) => (key === "DB_URL" ? dbUrl : dbToken),
+    );
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/turso-migration-lib.ts:209]: Resolved in 9382f63c. Replaced the ternary with an explicit lookup object `{ DB_URL: dbUrl, DB_TOKEN: dbToken }` so unknown keys return `undefined` instead of silently returning `dbToken`. `requiredEnv` then throws a clear error for any unexpected key, and there is no dead code branch to cover.

File: test/scripts/termination-signals.test.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  using _signal = stub(Deno, method, ((signal, listener) =>
+    calls.push([signal, listener])) as typeof Deno.addSignalListener);
+
+  listen(handler);
+
+  expect(calls).toEqual([
+    ["SIGINT", handler],
+    ["SIGTERM", handler],
+  ]);
+};
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/scripts/termination-signals.test.ts:24]: Resolved in 9382f63c. Added two tests: `continues when registering a signal listener throws` and `continues when removing a signal listener throws`. Both stub `Deno.addSignalListener`/`Deno.removeSignalListener` to throw, then assert `onTerminationSignals`/`offTerminationSignals` complete without propagating — covering the catch branch in `scripts/termination-signals.ts`.

File: src/shared/turso-api.ts [hunk 8]
```diff
@@ -1,98 +1,283 @@
...
+        `${failure}. Cleanup also failed: ${errorMessage(error)}`,
+      );
+    }
+  };
+
+  const createDatabase = async (
+    request: CreateTursoDatabaseRequest,
+  ): Promise<Result<TursoDatabaseCredentials>> => {
+    const name = slugifyForTurso(request.name);
+    const createResponse = await fetchApi(databasePath(request.organization), {
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/turso-api.ts]: Resolved in 9382f63c. The create POST is now inside a try/catch that calls `cleanUpFailedCreate` on any thrown error (abort, timeout, network), so an ambiguously accepted database is deleted. Merged the POST and parse steps into one try/catch to avoid duplication. The regression test `deletes the destination when create throws after accepting` locks the cleanup path.

File: scripts/database-snapshot-lib.ts [hunk 9]
```diff
@@ -308,9 +345,9 @@ export const createDatabaseSnapshot = async (
   return await withCleanup(async () => {
     const temporaryPath = join(temporaryDirectory, "snapshot.sqlite");
     writeProgress(SNAPSHOT_PROGRESS.syncing);
-    await syncReplica(temporaryPath, request, factory);
+    await syncReplica(temporaryPath, request, factory, signal);
     writeProgress(SNAPSHOT_PROGRESS.verifying);
-    await checkpointAndVerify(temporaryPath, factory);
+    await checkLocalSnapshot(temporaryPath, factory, SNAPSHOT_QUERY_CHECKS);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/database-snapshot-lib.ts]: Resolved in 9382f63c. `checkLocalSnapshot` now accepts an optional `AbortSignal` and races each `client.execute` call against it via `waitForOrAbort`, the same helper used for `syncReplica`. `createDatabaseSnapshot` forwards its signal to the verification call. The regression test `stops verification when its signal is aborted` reproduces a stalled `PRAGMA integrity_check` and asserts the client closes and temporary files are removed.

File: test/scripts/termination-signals.test.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  });
+  onTerminationSignals(() => {});
+});
+
+test("continues when removing a signal listener throws", () => {
+  using _signal = stub(Deno, "removeSignalListener", () => {
+    throw new Error("unsupported signal");
+  });
+  offTerminationSignals(() => {});
+});
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/scripts/termination-signals.test.ts]: Resolved in 1e402313. Updated both tests to throw only for SIGINT, track all calls, and assert SIGTERM was still attempted after the SIGINT failure. `continues to SIGTERM after SIGINT registration throws` and `continues to SIGTERM after SIGINT removal throws` now verify per-signal continuation rather than a loop-wide catch.

File: src/shared/turso-api.ts [hunk 11]
```diff
@@ -1,98 +1,288 @@
...
+        CreateTursoDatabaseSchema,
+        createResponse.text,
+        "Create database",
+      ).database;
+    } catch (error) {
+      return await cleanUpFailedCreate(
+        request.organization,
+        name,
+        `Create database failed: ${errorMessage(error)}`,
+      );
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/turso-api.ts:198]: Resolved in 1e402313. Split the try/catch: the create POST throw now returns `errorResult` without attempting cleanup, since we cannot prove the database was created. Only parse failures (where we have a response proving acceptance) trigger `cleanUpFailedCreate`. The test `does not delete when the create POST throws before a response` locks this — it asserts no DELETE request was made.

File: scripts/turso-migration-lib.ts [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+      );
+      deps.stderr(`Remove this directory: ${outcome.tempDirectory}`);
+      return 1;
+    }
+    return 0;
+  } catch (error) {
+    if (deps.signal.aborted) {
+      deps.stderr("Migration interrupted.");
+      return 130;
+    }
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/turso-migration-lib.ts:265]: Resolved in 1e402313. The interrupt handler now checks whether the error message contains cleanup info. If the error is just the abort reason (production path: `new Error("Migration interrupted")`), it shows the short message. If cleanup failures are embedded in the error, it shows the full message so the operator sees both the interruption and the cleanup problem. Two tests cover both branches: `cleans up remote and local state after an interrupted upload` (clean interrupt) and `reports cleanup failures during an interruption` (cleanup failure during interrupt).

File: src/shared/turso-api.ts [hunk 13]
```diff
@@ -1,98 +1,292 @@
...
+    }
+    if (database.Name !== name) {
+      return await cleanUpFailedCreate(
+        request.organization,
+        name,
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/turso-api.ts]: Resolved in 0366994f. The name-mismatch branch no longer calls `cleanUpFailedCreate`. It returns `errorResult` without deleting anything, since a mismatched response cannot prove which database was created. The test `does not delete on a name mismatch from the create response` asserts no DELETE request was made.

File: scripts/turso-migration-lib.ts [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+    deps.signal.throwIfAborted();
+    deps.stdout("Downloading the source database...");
+    const path = await deps.createSnapshot(
+      { ...source, outputPath: join(tempDirectory, "database.sqlite") },
+      deps.stdout,
+      deps.signal,
+    );
+    deps.signal.throwIfAborted();
+    deps.stdout("Checking the SQLite file for Turso...");
+    await deps.verifyUploadFile(path);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/turso-migration-lib.ts]: Resolved in 0366994f. `verifyTursoUploadFile` now accepts an optional `AbortSignal` and forwards it to `checkLocalSnapshot`, which races each `client.execute` call against the signal via `waitForOrAbort`. The migration passes `deps.signal` to the verify call.

File: scripts/turso-migration-file.ts [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+  const file = createReadStream(path);
+  const fileClosed = Promise.withResolvers<void>();
+  file.once("close", fileClosed.resolve);
+  request.once("error", file.destroy.bind(file));
+  request.once("error", response.reject);
+  file.once("error", request.destroy.bind(request));
+  file.once("error", response.reject);
+  file.pipe(request);
+  try {
+    return await response.promise;
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply scripts/turso-migration-file.ts:140]: No code change needed. Per AGENTS.md: "Do not add fallbacks, placeholders, or try/catches for failures that can only happen when a foundational system is already broken." HTTP/1.1 servers with `Content-Length` cannot send a response before the full request body is received — the server knows exactly how many bytes to expect. An early 2xx before the file stream finishes would be an HTTP protocol violation. I attempted to wait for the request `finish` event, but it caused hangs in the Deno Node compat layer because `finish` does not fire reliably on `ClientRequest` when the server closes the connection. Defending against this edge case would introduce fragile event-watching code for a condition that cannot occur with a compliant HTTP server.

File: test/test-utils/turso-migration.ts [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+    stdout: (line) => state.stdout.push(line),
+    uploadDatabaseFile: (path) => {
+      state.events.push("upload");
+      state.uploads.push(path);
+      return Promise.resolve();
+    },
+    verifyUploadFile: (_path: string, _signal: AbortSignal) => {
+      state.events.push("verify");
+      return Promise.resolve();
+    },
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply test/test-utils/turso-migration.ts]: Resolved in ae7a1d54. The test fake now records `signal` in `state.verifySignals`, and the happy-path test `downloads, verifies, and uploads the database file` asserts `state.verifySignals` equals `[state.deps.signal]`.
